python/dgl/data/movielens.py

```diff
@@ -0,0 +1,644 @@
+"""MovieLens dataset"""
+import os
+import zipfile
+import shutil
+
+import numpy as np
+import pandas as pd
+
+
+from torch import LongTensor, Tensor
+from ..convert import graph
+from ..transforms.functional import add_reverse_edges
+from .dgl_dataset import DGLDataset
+
+from .utils import (
+    download,
+    load_graphs,
+    load_info,
+    save_graphs,
+    save_info,
+    split_dataset,
+)
+
+GENRES_ML_100K = [
+    "unknown",
+    "Action",
+    "Adventure",
+    "Animation",
+    "Children",
+    "Comedy",
+    "Crime",
+    "Documentary",
+    "Drama",
+    "Fantasy",
+    "Film-Noir",
+    "Horror",
+    "Musical",
+    "Mystery",
+    "Romance",
+    "Sci-Fi",
+    "Thriller",
+    "War",
+    "Western",
+]
+GENRES_ML_1M = GENRES_ML_100K[1:]
+GENRES_ML_10M = GENRES_ML_100K + ["IMAX"]
+
+try:
+    import torch
+except ImportError:
+    HAS_TORCH = False
+else:
+    HAS_TORCH = True
+
+def check_pytorch():
+    """Check if PyTorch is the backend."""
+    if not HAS_TORCH:
+        raise ModuleNotFoundError(
+            "MovieLensDataset requires PyTorch to be the backend."
+        )
+
+class MovieLensDataset(DGLDataset):
+    r"""MovieLens dataset for edge prediction tasks. The raw datasets are extracted from 
+    `MovieLens <https://grouplens.org/datasets/movielens/>`, introduced by
+    `Movielens unplugged: experiences with an occasionally connected recommender system <https://dl.acm.org/doi/10.1145/604045.604094>`.
+
+    The datasets consist of user ratings for movies and incorporate additional user/movie information in the form of features. 
+    The nodes represent users and movies, and the edges store ratings that users assign to movies.
+
+    Statistics:
+
+    MovieLens-100K (ml-100k)
+
+    - Users: 943
+    - Movies: 1,682
+    - Ratings: 100,000 (1, 2, 3, 4, 5)
+
+    MovieLens-1M (ml-1m)
+
+    - Users: 6,040
+    - Movies: 3,706
+    - Ratings: 1,000,209 (1, 2, 3, 4, 5)
+
+    MovieLens-10M (ml-10m)
+
+    - Users: 69,878 
+    - Movies: 10,677
+    - Ratings: 10,000,054 (0.5, 1, 1.5, ..., 4.5, 5.0)
+
+    Parameters
+    ----------
+    name: str
+        Dataset name. (:obj:`"ml-100k"`, :obj:`"ml-1m"`, :obj:`"ml-10m"`). 
+    valid_ratio: int
+        Ratio of validation samples out of the whole dataset. Should be in (0.0, 1.0). 
+    test_ratio: int, optional
+        Ratio of testing samples out of the whole dataset. Should be in (0.0, 1.0) for :obj:`"ml-1m"` and :obj:`"ml-10m"`. This parameter is invalid
+        when :obj:`name` is :obj:`"ml-100k"`, since its testing samples are pre-specified. 
+        Default: None
+    raw_dir : str, optional
+        Raw file directory to download/store the data.
+        Default: ~/.dgl/
+    force_reload : bool, optional
+        Whether to re-download(if the dataset has not been downloaded) and re-process the dataset. 
```

Comment:
"(if the dataset has not been downloaded)" this is not needed since the dataset will be downloaded in that case anyway

this is more about if the dataset on s3 has been updated

Reply:
Sorry but I don't get why the dataset will be re-downloaded with force_reload if the raw datasets have already been downloaded in the local storage? I have checked codes in `dgl_dataset.py` and see there are logics checking if there are existing datasets in local before downloading them with force_reload.

Reply:
And could you elaborate "_this is more about if the dataset on s3 has been updated_"? I did't get this point.

Reply:
> Sorry but I don't get why the dataset will be re-downloaded with force_reload if the raw datasets have already been downloaded in the local storage? I have checked codes in dgl_dataset.py and see there are logics checking if there are existing datasets in local before downloading them with force_reload.

This is not right. See https://github.com/dmlc/dgl/blob/master/python/dgl/data/dgl_dataset.py#L186. `force_reload` will force re-downloading.

Reply:
> And could you elaborate "this is more about if the dataset on s3 has been updated"? I did't get this point.

It's possible that the url stays unchanged, but the data source corresponding to the url has been changed and updated. In that case, it's necessary to force re-downloading. This is rare though.

Reply:
> > Sorry but I don't get why the dataset will be re-downloaded with force_reload if the raw datasets have already been downloaded in the local storage? I have checked codes in dgl_dataset.py and see there are logics checking if there are existing datasets in local before downloading them with force_reload.
> 
> This is not right. See https://github.com/dmlc/dgl/blob/master/python/dgl/data/dgl_dataset.py#L186. `force_reload` will force re-downloading.

But look at the implementations of `self._download()`, https://github.com/dmlc/dgl/blob/4c01563245c46aadcd35d0869a8fb5d7962408d2/python/dgl/data/dgl_dataset.py#L159, dgll will test whether dataset has been downloaded first. If yes, then return.

Reply:
My bad. You are right. Then it always requires a new URL as raw_path has a suffix based on the URL.